File: src/main/java/com/nickmcdowall/lsd/interceptor/rest/LsdRestTemplateCustomizer.java
package com.nickmcdowall.lsd.interceptor.rest;

import lombok.RequiredArgsConstructor;
import lombok.Value;
import org.springframework.boot.web.client.RestTemplateCustomizer;
import org.springframework.http.client.BufferingClientHttpRequestFactory;
import org.springframework.http.client.ClientHttpRequestInterceptor;
import org.springframework.http.client.SimpleClientHttpRequestFactory;
import org.springframework.web.client.RestTemplate;

import java.util.List;

@Value
@RequiredArgsConstructor
public class LsdRestTemplateCustomizer implements RestTemplateCustomizer {

    ClientHttpRequestInterceptor interceptor;

    @Override
    public void customize(RestTemplate restTemplate) {
        List<ClientHttpRequestInterceptor> interceptors = restTemplate.getInterceptors();
        if (!interceptors.contains(interceptor)) {
            interceptors.add(interceptor);
        }
        restTemplate.setRequestFactory(new BufferingClientHttpRequestFactory(new SimpleClientHttpRequestFactory()));
    }
}

File: lsd-interceptors/src/main/kotlin/io/lsdconsulting/interceptors/http/LsdRestTemplateCustomizer.kt
package io.lsdconsulting.interceptors.http

import org.springframework.boot.web.client.RestTemplateCustomizer
import org.springframework.http.client.BufferingClientHttpRequestFactory
import org.springframework.http.client.ClientHttpRequestInterceptor
import org.springframework.http.client.SimpleClientHttpRequestFactory
import org.springframework.web.client.RestTemplate

data class LsdRestTemplateCustomizer(
    private var interceptor: ClientHttpRequestInterceptor
) : RestTemplateCustomizer {

    override fun customize(restTemplate: RestTemplate) {
        val interceptors = restTemplate.interceptors
        if (!interceptors.contains(interceptor)) {
            interceptors.add(interceptor)
        }
        restTemplate.requestFactory = BufferingClientHttpRequestFactory(SimpleClientHttpRequestFactory())
    }
}
